### PR TITLE
Configurable query parameters parser

### PR DESCRIPTION
**Pull request by wp0pw**
Implementing changes for  #1128 
Added new mux Option to configure query parameter parser to use with instance of mux.

### CODE REVIEW

File: runtime/mux.go [hunk 1]
```diff
@@ -55,6 +58,19 @@ func WithForwardResponseOption(forwardResponseOption func(context.Context, http.
 	}
 }
 
+// A QueryParameterParser populates message form HTTP query parameters values
+type QueryParameterParser func(msg proto.Message, values url.Values, filter *utilities.DoubleArray) error
```
Comment (johanbrandhorst) [inline runtime/mux.go]: What do you think about making this an interface?

File: runtime/mux.go [hunk 2]
```diff
@@ -55,6 +58,19 @@ func WithForwardResponseOption(forwardResponseOption func(context.Context, http.
 	}
 }
 
+// A QueryParameterParser populates message form HTTP query parameters values
```
Comment (johanbrandhorst) [inline runtime/mux.go]:
```suggestion
// A QueryParameterParser populates the message from HTTP query parameter values
```

File: runtime/mux.go [hunk 3]
```diff
@@ -55,6 +58,19 @@ func WithForwardResponseOption(forwardResponseOption func(context.Context, http.
...
+type QueryParameterParser func(msg proto.Message, values url.Values, filter *utilities.DoubleArray) error
+
+// WithQueryParametersPopulator returns a ServeMuxOption representing new populator used to populate message from query parameters
```
Comment (johanbrandhorst) [inline runtime/mux.go]:
```suggestion
// WithQueryParameterParser sets the query parameter parser, used to populate message from query parameters, on this ServeMux.
```

File: runtime/mux.go [hunk 4]
```diff
@@ -55,6 +58,19 @@ func WithForwardResponseOption(forwardResponseOption func(context.Context, http.
...
+//
+// queryParametersPopulator is an option that will be called on the proto.Message
+// url.Values, and utilities.DoubleArray on incoming requests.
```
Comment (johanbrandhorst) [inline runtime/mux.go]: This isn't necessary

File: runtime/mux.go [hunk 5]
```diff
@@ -55,6 +58,19 @@ func WithForwardResponseOption(forwardResponseOption func(context.Context, http.
...
+func WithQueryParametersPopulator(queryParametersPopulator QueryParameterParser) ServeMuxOption {
```
Comment (johanbrandhorst) [inline runtime/mux.go]: Lets name this more consistently:
```suggestion
func WithQueryParameterParser(queryParameterParser QueryParameterParser) ServeMuxOption {
```

File: runtime/mux.go [hunk 6]
```diff
@@ -55,6 +58,19 @@ func WithForwardResponseOption(forwardResponseOption func(context.Context, http.
...
+	return func(serveMux *ServeMux) {
+		serveMux.queryParameterParser = queryParametersPopulator
```
Comment (johanbrandhorst) [inline runtime/mux.go]:
```suggestion
		serveMux.queryParameterParser = queryParameterParser
```

**Comment by wp0pw** [comment]
Global parser is fine I think, if we don't want to change the generated code there is not much else we can do here - I'll try to make changes some time this week, thanks for quick feedback.

File: runtime/mux.go (hunk 1 above)
Comment (wp0pw) [reply runtime/mux.go]: can do, will make the changes for global parser and see if interface fits/is needed.

File: runtime/mux.go (hunk 4 above)
Comment (wp0pw) [reply runtime/mux.go]: removed

**Comment by wp0pw** [comment]
Changed to use global parser. Added interface for parser implementations.
No changes to generated code.

File: runtime/mux.go (hunk 1 above)
Comment (wp0pw) [reply runtime/mux.go]: added interface in query.go

File: runtime/query.go [hunk 7]
```diff
@@ -17,9 +17,25 @@ import (
...
+}
+
+// PopulateQueryParameters parses query parameters
+// into "msg" using current query parser
 func PopulateQueryParameters(msg proto.Message, values url.Values, filter *utilities.DoubleArray) error {
+	return currentQueryParser.Parse(msg, values, filter)
+}
+
+// DefaultQueryParser is a default implementation of QueryParametersParser
+type DefaultQueryParser struct{}
```
[the remark this answers is not in the document]
Comment (wp0pw) [reply runtime/query.go]: unexported

File: runtime/mux.go [hunk 8]
```diff
@@ -55,6 +55,13 @@ func WithForwardResponseOption(forwardResponseOption func(context.Context, http.
 	}
 }
 
+// WithQueryParameterParser sets the query parameter parser, used to populate message from query parameters.
+func WithQueryParameterParser(queryParameterParser QueryParameterParser) ServeMuxOption {
```
[the remark this answers is not in the document]
Comment (wp0pw) [reply runtime/mux.go]: it's a `Set` now

**Comment by wp0pw** [comment]
Pleasure!

**Comment by anilCSE** [comment]
Is there any example implementation for this? I was looking at the docs, didn't find any info

**Comment by johanbrandhorst** [comment]
The default implementation in both master and v2 implements the interface. You'll just have to design your own with the same signature. See https://github.com/grpc-ecosystem/grpc-gateway/blob/master/runtime/query.go#L37

**Comment by anilCSE** [comment]
Looks like there's some issue with `gogoproto` support. 
```
 cannot use &customQueryParser literal (type *customQueryParser) as type runtime.QueryParameterParser in assignment:
        *customQueryParser does not implement runtime.QueryParameterParser (wrong type for Parse method)
                have Parse("github.com/gogo/protobuf/proto".Message, url.Values, *utilities.DoubleArray) error
                want Parse(protoiface.MessageV1, url.Values, *utilities.DoubleArray) error
```

Any known workarounds for this?

**Comment by johanbrandhorst** [comment]
No, you can't use it with Gogo protobuf

**Comment by anilCSE** [comment]
Oh okay, thank you for the info @johanbrandhorst . We are trying to `convert` query parameters to grpc server specific types in `grpc-gateway`. Is there an easy way to achieve this? For the sake of `UX`, we want to keep `human-readable` parameters for clients and convert them to `bytes` in `grpc-gateway`'s middleware/interceptor (?).

**Comment by johanbrandhorst** [comment]
Any parameters not captured in the body or the path automatically become query parameters, are you sure you need to do anything?

**Comment by anilCSE** [comment]
I need to cast them before forwarding them to grpc server. We have some prefix encoded strings for clients and grpc server uses bytes (proto definitions are bytes). We want to support strings from client and cast them to bytes in grpc gateway middleware

**Comment by achew22** [comment]
@anilCSE you should be able to write anything you want in some kind of Go middleware. grpc-gateway doesn't have any provisions for doing this directly so you will have to write something custom up.

**Pull request merged**